core/isolate.rs

```diff
@@ -511,9 +511,11 @@ impl Isolate {
         let op_id = 0;
         Some((op_id, buf))
       }
-      Op::Async(fut) => {
+      Op::Async(fut, blocks_exit) => {
         let fut2 = fut.map_ok(move |buf| (op_id, buf));
-        self.pending_ops.push(fut2.boxed());
+        self
+          .pending_ops
+          .push(Pin::new(Box::new(PendingOp(fut2.boxed(), blocks_exit))));
         self.have_unpolled_ops = true;
```

Comment:
I'd recommend using `Op::Async` and `Op::AsyncOptional`.

`Op::AsyncOptional` should be kept in `Isolate.pending_optional_ops` - which is the same type as `Isolate.pending_ops`.

Reply:
I created another version of this at #3721, which uses `AsyncOptional` instead of adding flag to `Async`. I'd like to avoid having `Isolate.pending_optional_ops` because if we have 2 `FuturesUnordered` list in isolate, it seems to me that it's difficult to poll these 2 futures list correctly.